OpenRA.Mods.AS/Projectiles/CyclicImpactProjectile.cs

```diff
@@ -0,0 +1,262 @@
+		int lifespan;
+		int ticks;
+		int mindelay;
+		World world;
+		WRot offsetRotation;
+		CyclicImpactProjectileEffect[] projectiles; // offset projectiles
+
+		public Actor SourceActor { get { return args.SourceActor; } }
+
+		public CyclicImpactProjectile(CyclicImpactProjectileInfo info, ProjectileArgs args)
+		{
+			this.info = info;
+			this.args = args;
+
+			projectilepos = args.Source;
+			sourcepos = args.Source;
+
+			var firedBy = args.SourceActor;
+
+			world = args.SourceActor.World;
+
+			if (info.Speed.Length > 1)
+				speed = new WDist(world.SharedRandom.Next(info.Speed[0].Length, info.Speed[1].Length));
+			else
+				speed = info.Speed[0];
+
+			targetpos = GetTargetPos();
+
+			mindelay = args.Weapon.MinRange.Length / speed.Length;
+
+			projectiles = new CyclicImpactProjectileEffect[info.Offsets.Count()];
+			var range = Util.ApplyPercentageModifiers(args.Weapon.Range.Length, args.RangeModifiers);
+			var mainfacing = (targetpos - sourcepos).Yaw.Facing;
+
+			// target that will be assigned
+			Target target;
+
+			// main bullet facing
+			int facing = 0;
+
+			for (int i = 0; i < info.Offsets.Count(); i++)
+			{
+				switch (info.FireMode)
+				{
+					case FireMode.Focus:
+						offsetRotation = WRot.FromFacing(mainfacing);
+						offsetTargetPos = sourcepos + new WVec(range, 0, 0).Rotate(offsetRotation);
+						offsetSourcePos = sourcepos + info.Offsets[i].Rotate(offsetRotation);
+						break;
+					case FireMode.Line:
+						offsetRotation = WRot.FromFacing(mainfacing);
+						offsetTargetPos = sourcepos + new WVec(range + info.Offsets[i].X, info.Offsets[i].Y, info.Offsets[i].Z).Rotate(offsetRotation);
+						offsetSourcePos = sourcepos + info.Offsets[i].Rotate(offsetRotation);
+						break;
+					case FireMode.Spread:
+						offsetRotation = WRot.FromFacing(info.Offsets[i].Yaw.Facing - 64) + WRot.FromFacing(mainfacing);
+						offsetSourcePos = sourcepos + info.Offsets[i].Rotate(offsetRotation);
+						offsetTargetPos = sourcepos + new WVec(range + info.Offsets[i].X, info.Offsets[i].Y, info.Offsets[i].Z).Rotate(offsetRotation);
+						break;
+				}
+
+				if (info.Inaccuracy.Length > 0)
+				{
+					var inaccuracy = Util.ApplyPercentageModifiers(info.Inaccuracy.Length, args.InaccuracyModifiers);
+					var maxOffset = inaccuracy * (args.PassiveTarget - projectilepos).Length / range;
+					var inaccoffset = WVec.FromPDF(world.SharedRandom, 2) * maxOffset / 1024;
+					offsetTargetPos += inaccoffset;
+				}
+
+				target = Target.FromPos(offsetTargetPos);
+
+				var estmatedlifespan = Math.Max(args.Weapon.Range.Length / speed.Length, 1);
+
+				// if it's true then lifespan is counted from source pos to target, instead of max range
+				lifespan = info.KillProjectilesWhenReachedTargetLocation
+					? Math.Max((args.PassiveTarget - args.Source).Length / speed.Length, 1)
+					: estmatedlifespan;
+
+				facing = (offsetTargetPos - offsetSourcePos).Yaw.Facing;
+				var pargs = new ProjectileArgs
```

Comment:
Don't shorten this either, this sounds like some monster from the Star Wars Expanded Universe. :D